feat: command to add an asset to an item

sunu opened the pull request:
**Related Issue(s):**
https://github.com/stac-utils/stactools/issues/147

**Description:**
Adds a command called `addasset` to add an asset to an item. It takes an item's path, an asset's key, path and optionally roles, type etc as argument. The command doesn't handle `extra_fields` yet. 

**PR checklist:**

- [x] Code is formatted (run `scripts/format`).
- [x] Code lints properly (run `scripts/lint`).
- [x] Tests pass (run `scripts/test`).
- [x] Documentation has been updated to reflect changes, if applicable.
- [x] Changes are added to the [CHANGELOG](../CHANGELOG.md).

sunu commented:
The failing tests on Python 3.10 seem to be unrelated to this PR. I've created a separate PR (https://github.com/stac-utils/stactools/pull/301) to handle that.

gadomski commented on tests/cli/commands/test_addasset.py:
```diff
@@ -0,0 +1,113 @@
+
+    def test_add_asset_to_item(self):
+        with TemporaryDirectory() as tmp_dir:
+            catalog = create_temp_catalog_copy(tmp_dir)
```
I believe the `catalog` part of this test is not necessary, since we are just testing the behavior on the `Item`. The test should be reduced to only include the components necessary to test the expected behavior.

gadomski commented on src/stactools/core/addasset.py:
```diff
@@ -0,0 +1,57 @@
+            This operation mutates the Item.
+    """
+    item_href = item.get_self_href()
+    if item_href is None:
+        raise ValueError(f"Target Item {item.id} must have an HREF to add an asset")
```
If `move_assets` is `False` and the asset href is absolute, I don't see any reason to error if the item doesn't have an href.

sunu commented on src/stactools/cli/commands/addasset.py:
```diff
@@ -0,0 +1,89 @@
+        "--ignore-conflicts",
+        is_flag=True,
+        help=(
+            "If there is a conflict with an existing asset, do not raise an "
+            "error, leave the original asset from the target item in place."
```
Yup, that's exactly what it means. I'll update description to be clearer.

sunu commented on src/stactools/core/addasset.py:
```diff
@@ -0,0 +1,57 @@
+    else:
+        asset_href = asset.get_absolute_href()
+        if asset_href is None:
+            raise ValueError(f"Asset {asset.title} must have an HREF to be added")
```
That's true. I think using `repr(asset)` would be better.

sunu commented on tests/cli/commands/test_addasset.py:
```diff
@@ -0,0 +1,113 @@
+    return temp_path
+
+
+class AddAssetTest(CliTestCase):
```
Yes, makes sense

sunu commented on src/stactools/cli/commands/addasset.py:
```diff
@@ -0,0 +1,89 @@
+    ignore_conflicts: bool = False,
+) -> None:
+    item = pystac.read_file(item_path)
+    if not isinstance(item, pystac.Item):
```
I briefly looked into what's required to support both an `Item` or a `Collection` as the target. From what I understand, we would need an equivalent of the `move_asset_file_to_item` function for `Collections`: https://github.com/stac-utils/stactools/blob/1077ccd58cafc4dd392abddefc490af2ee64c1ea/src/stactools/core/copy.py#L14

How do you feel about opening an issue for that and adding support for `Collections` in a separate PR?

sunu commented:
Hi @gadomski, I think I have addressed most of your comments and this should be ready for another look. Also, thank you for taking the time to providing so much feedback. I'm pretty new to STAC. So I really appreciate it 🙂

gadomski commented on tests/core/test_addasset.py:
```diff
@@ -0,0 +1,74 @@
+            )
+            item = add_asset_to_item(item, "test-asset", asset)
+
+            asset = item.get_assets().get("test-asset")
```
```suggestion
            asset = item.assets["test-asset")]
```

gadomski commented on tests/core/test_addasset.py:
```diff
@@ -0,0 +1,74 @@
+
+class AddAssetTest(TestCase):
+    def test_add_asset_to_item(self):
+        with TemporaryDirectory() as tmp_dir:
+            item_path = create_temp_copy(
+                test_data.get_path("data-files/core/simple-item.json"),
+                tmp_dir,
+                "item.json",
+            )
```
Since we're not moving the asset, do we need to create a copy of the item in this unit test?

gadomski commented on src/stactools/core/addasset.py:
```diff
@@ -0,0 +1,62 @@
+            raise ValueError(
+                f"Asset {repr(asset)} must have an absolute HREF to be added"
+            )
+        if item_href is None and (move_assets or not is_absolute_href(asset.href)):
+            # An absolute href for the item is required if we need to move the
+            # asset into the item or if the asset href is relative
+            raise ValueError(
+                f"Target Item {repr(item)} must have an HREF to add an asset"
+            )
```
Syntax isn't perfect but you get the idea

```suggestion
        if item_href is None and move_assets:
            # An absolute href for the item is required if we need to move the
            # asset into the item or if the asset href is relative
            raise ValueError(
                f"Target Item {repr(item)} must have an HREF to move an asset to the item"
            )
        elif item_href is None and not is_absolute_href(asset.href):
            raise ValueError(f"Target Item {repr(item)} must have an href to add an asset with a relative href")
```

gadomski commented on tests/core/test_addasset.py:
```diff
@@ -0,0 +1,74 @@
+            assert asset.description == "placeholder asset", asset.to_dict()
+            self.assertListEqual(asset.roles, ["thumbnail", "overview"])
+
+    def add_asset_move(self):
```
```suggestion
    def test_add_asset_move(self):
```

Otherwise, it isn't picked up as a test.

sunu replied:
oops, I totally overlooked this one.

gadomski commented on tests/core/test_addasset.py:
```diff
@@ -0,0 +1,74 @@
+            self.assertListEqual(asset.roles, ["thumbnail", "overview"])
+
+    def add_asset_move(self):
+        with TemporaryDirectory() as tmp_dir:
+            item_path = create_temp_copy(
+                test_data.get_path("data-files/core/simple-item.json"),
+                tmp_dir,
+                "item.json",
+            )
+            item = pystac.read_file(item_path)
+            assert isinstance(item, pystac.Item)
+            asset = item.get_assets().get("test-asset")
+            assert asset is None
+            with TemporaryDirectory as tmp_dir2:
+                asset_path = create_temp_copy(
+                    test_data.get_path("data-files/core/byte.tif"), tmp_dir2, "test.tif"
+                )
+                asset = pystac.Asset(
+                    asset_path,
+                    "test",
+                    "placeholder asset",
+                    roles=["thumbnail", "overview"],
+                )
+                item = add_asset_to_item(
+                    item, "test-asset", asset, move_assets=True, ignore_conflicts=True
+                )
+
+                asset = item.get_assets().get("test-asset")
+                assert isinstance(asset, pystac.Asset), asset
+                assert asset.href is not None, asset.to_dict()
+                assert os.path.isfile(asset.href), asset.to_dict()
+                assert asset.title == "test", asset.to_dict()
+                assert asset.description == "placeholder asset", asset.to_dict()
+                self.assertEqual(
+                    os.path.dirname(asset.get_absolute_href()),
+                    os.path.dirname(item.get_self_href()),
+                )
```
This is a bit of a style thing, but a lot of these assertions are covered by `test_add_asset_to_item`, so don't need to be made twice. This could be simplified to something like:

```suggestion
        with TemporaryDirectory() as tmp_dir:
            item_path = create_temp_copy(
                test_data.get_path("data-files/core/simple-item.json"),
                tmp_dir,
                "item.json",
            )
            item = pystac.read_file(item_path)
            with TemporaryDirectory as tmp_dir2:
                asset_path = create_temp_copy(
                    test_data.get_path("data-files/core/byte.tif"), tmp_dir2, "test.tif"
                )
                asset = pystac.Asset(
                    asset_path,
                    "test",
                    "placeholder asset",
                    roles=["thumbnail", "overview"],
                )
                item = add_asset_to_item(
                    item, "test-asset", asset, move_assets=True, ignore_conflicts=True
                )

                asset = item.assets["test-asset"]
                assert isinstance(asset, pystac.Asset), asset
                assert asset.href is not None, asset.to_dict()
                assert os.path.isfile(asset.href), asset.to_dict()
                self.assertEqual(
                    os.path.dirname(asset.get_absolute_href()),
                    os.path.dirname(item.get_self_href()),
                )
```

gadomski commented on tests/core/test_addasset.py:
```diff
@@ -0,0 +1,74 @@
+from tests.utils import create_temp_copy
+
+
+class AddAssetTest(TestCase):
```
Thanks for adding some `core` unit tests! I think there are a couple more cases that would be useful to assert:

1. Asserting that an error is raised if `move=True` and the item doesn't have an href
2. Asserting that things work ok if the item doesn't have an href, `move=False`, and the asset href is absolute
3. Asserting an error if the item doesn't have an href, `move=False`, and the asset href is relative

There may be a couple others. When reading your tests, I was curious what happened in the above cases, and had to go read the code to find out.

sunu replied:
I have added tests for these cases and a few more.

philvarner commented on src/stactools/cli/commands/addasset.py:
```diff
@@ -0,0 +1,92 @@
+
+
+def create_addasset_command(cli: click.Group) -> click.Command:
+    @cli.command("addasset", short_help="Add an asset to an item.")
```
We should establish a convention for multi-word commands. We currently only have two cases: `create-item` and `addraster`. I think we should standardize on kebab-case, and change addraster. @gadomski

gadomski replied:
👍🏽 to kebab case, ticket created: https://github.com/stac-utils/stactools/issues/322.

philvarner commented on src/stactools/core/addasset.py:
```diff
@@ -0,0 +1,62 @@
+    if key in item.assets:
+        if not ignore_conflicts:
+            raise Exception(
+                "Target item {} already has asset with key {}, "
```
I would prefer f-string use here rather than format

philvarner commented on src/stactools/core/addasset.py:
```diff
@@ -0,0 +1,62 @@
+                "cannot add asset in from {}".format(item, key, asset_href)
+            )
+    else:
+        if asset_href is None:
```
```suggestion
        if not asset_href:
```

to also handle the case where asset_href is the empty string

sunu commented:
@gadomski @philvarner I have updated the code to address the review comments.

gadomski commented on src/stactools/cli/commands/add_asset.py:
```diff
@@ -0,0 +1,92 @@
+        ignore_conflicts: bool = False,
+    ) -> None:
+        _add_asset(
+            item_path,
+            asset_key,
+            asset_path,
```
```suggestion
            pystac.utils.make_absolute_href(item_path),
            asset_key,
            pystac.utils.make_absolute_href(asset_path),
```

gadomski commented on src/stactools/core/add_asset.py:
```diff
@@ -0,0 +1,66 @@
+            This operation mutates the Item.
+    """
+    item_href = item.get_self_href()
+    asset_href = asset.get_absolute_href()
```
```suggestion
    asset_href = asset.href
```

gadomski reviewed (changes requested):
I tried the following command and got an error:

```shell
$ stac add-asset tests/data-files/core/simple-item.json foobar tests/data-files/core/byte.tif
Traceback (most recent call last):
  <snip>
ValueError: Asset <Asset href=tests/data-files/core/byte.tif> must have an href to be added. The href value should be an absolute path or URL.
```

I think my proposed use-case (two relative, local hrefs) should be supported, so I've suggested changes that fix the error. Let me know what you think.

gadomski commented:
I've rebased this branch onto **main** and consolidated @sunu's contributions to one commit.

sunu commented:
Thank you for finishing the PR @gadomski! Sorry for dropping the ball in the middle.

gadomski commented:
No problem, thanks for the contribution!
